TypeScript thinks that primitives are Objects

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.1.0-dev.20180807

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
typescript number instanceof Object
typescript string instanceof Object
typescript primitive instanceof Object
typescript thinks strings are objects
typescript thinks numbers are objects

The results are the same in v2.9.2, so I am shocked that I didn't find any other reports of this issue.

**Code** (Tested with default config from `tsc --init`)

```ts
function distinguish(thing: string|number|Date) {
  if (thing instanceof Object) {
    console.log("Aha!! It's a Date in " + thing.getFullYear());
  } else if (typeof thing === 'string') {
    console.log("Aha!! It's a string of length " + thing.length);
  } else {
    console.log("Aha!! It's the number " + thing.toPrecision(3));
  }
}

distinguish(new Date());
distinguish("beef");
distinguish(3.14159265);
```

**Expected behavior:**

In JavaScript, `23 instanceof Object` and `"foo" instanceof Object` are both false, so this code works and should compile without errors.

**Actual behavior:**

`tsc` decides that `thing instanceof Object` is always true and gives three errors:

~~~
weird.ts:3:49 - error TS2339: Property 'getFullYear' does not exist on type 'string | number | Date'.
  Property 'getFullYear' does not exist on type 'string'.

3     console.log("Aha!! It's a Date in " + thing.getFullYear());
                                                  ~~~~~~~~~~~

weird.ts:5:58 - error TS2339: Property 'length' does not exist on type 'never'.

5     console.log("Aha!! It's a string of length " + thing.length);
                                                           ~~~~~~

weird.ts:7:50 - error TS2339: Property 'toPrecision' does not exist on type 'never'.

7     console.log("Aha!! It's the number " + thing.toPrecision(3));
                                                   ~~~~~~~~~~~
~~~

[**Playground Link**](https://www.typescriptlang.org/play/#src=function%20distinguish(thing%3A%20string%7Cnumber%7CDate)%20%7B%0D%0A%20%20if%20(thing%20instanceof%20Object)%20%7B%0D%0A%20%20%20%20console.log(%22Aha!!%20It's%20a%20Date%20in%20%22%20%2B%20thing.getFullYear())%3B%0D%0A%20%20%7D%20else%20if%20(typeof%20thing%20%3D%3D%3D%20'string')%20%7B%0D%0A%20%20%20%20console.log(%22Aha!!%20It's%20a%20string%20of%20length%20%22%20%2B%20thing.length)%3B%0D%0A%20%20%7D%20else%20%7B%0D%0A%20%20%20%20console.log(%22Aha!!%20It's%20the%20number%20%22%20%2B%20thing.toPrecision(3))%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Adistinguish(new%20Date())%3B%0D%0Adistinguish(%22beef%22)%3B%0D%0Adistinguish(3.14159265)%3B%0D%0A) <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->

**Related Issues:** <!-- Did you find other bugs that looked similar? -->

Possibly related to #13327

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -10595,6 +10595,7 @@ namespace ts {
         function isTypeDerivedFrom(source: Type, target: Type): boolean {
             return source.flags & TypeFlags.Union ? every((<UnionType>source).types, t => isTypeDerivedFrom(t, target)) :
                 target.flags & TypeFlags.Union ? some((<UnionType>target).types, t => isTypeDerivedFrom(source, t)) :
+                source.flags & TypeFlags.Primitive && !(target.flags & TypeFlags.Primitive) ? false :
                 source.flags & TypeFlags.InstantiableNonPrimitive ? isTypeDerivedFrom(getBaseConstraintOfType(source) || emptyObjectType, target) :
                 target === globalObjectType || target === globalFunctionType ? isTypeSubtypeOf(source, target) :
                 hasBaseType(source, getTargetType(target));

--- a/tests/baselines/reference/controlFlowInstanceOfGuardPrimitives.js
+++ b/tests/baselines/reference/controlFlowInstanceOfGuardPrimitives.js
@@ -1,0 +1,30 @@
+//// [controlFlowInstanceOfGuardPrimitives.ts]
+function distinguish(thing: string | number | Date) {
+    if (thing instanceof Object) {
+        console.log("Aha!! It's a Date in " + thing.getFullYear());
+    } else if (typeof thing === 'string') {
+        console.log("Aha!! It's a string of length " + thing.length);
+    } else {
+        console.log("Aha!! It's the number " + thing.toPrecision(3));
+    }
+}
+
+distinguish(new Date());
+distinguish("beef");
+distinguish(3.14159265);
+
+//// [controlFlowInstanceOfGuardPrimitives.js]
+function distinguish(thing) {
+    if (thing instanceof Object) {
+        console.log("Aha!! It's a Date in " + thing.getFullYear());
+    }
+    else if (typeof thing === 'string') {
+        console.log("Aha!! It's a string of length " + thing.length);
+    }
+    else {
+        console.log("Aha!! It's the number " + thing.toPrecision(3));
+    }
+}
+distinguish(new Date());
+distinguish("beef");
+distinguish(3.14159265);

--- a/tests/baselines/reference/controlFlowInstanceOfGuardPrimitives.symbols
+++ b/tests/baselines/reference/controlFlowInstanceOfGuardPrimitives.symbols
@@ -1,0 +1,50 @@
+=== tests/cases/conformance/controlFlow/controlFlowInstanceOfGuardPrimitives.ts ===
+function distinguish(thing: string | number | Date) {
+>distinguish : Symbol(distinguish, Decl(controlFlowInstanceOfGuardPrimitives.ts, 0, 0))
+>thing : Symbol(thing, Decl(controlFlowInstanceOfGuardPrimitives.ts, 0, 21))
+>Date : Symbol(Date, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --), Decl(lib.scripthost.d.ts, --, --))
+
+    if (thing instanceof Object) {
+>thing : Symbol(thing, Decl(controlFlowInstanceOfGuardPrimitives.ts, 0, 21))
+>Object : Symbol(Object, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --))
+
+        console.log("Aha!! It's a Date in " + thing.getFullYear());
+>console.log : Symbol(Console.log, Decl(lib.dom.d.ts, --, --))
+>console : Symbol(console, Decl(lib.dom.d.ts, --, --))
+>log : Symbol(Console.log, Decl(lib.dom.d.ts, --, --))
+>thing.getFullYear : Symbol(Date.getFullYear, Decl(lib.es5.d.ts, --, --))
+>thing : Symbol(thing, Decl(controlFlowInstanceOfGuardPrimitives.ts, 0, 21))
+>getFullYear : Symbol(Date.getFullYear, Decl(lib.es5.d.ts, --, --))
+
+    } else if (typeof thing === 'string') {
+>thing : Symbol(thing, Decl(controlFlowInstanceOfGuardPrimitives.ts, 0, 21))
+
+        console.log("Aha!! It's a string of length " + thing.length);
+>console.log : Symbol(Console.log, Decl(lib.dom.d.ts, --, --))
+>console : Symbol(console, Decl(lib.dom.d.ts, --, --))
+>log : Symbol(Console.log, Decl(lib.dom.d.ts, --, --))
+>thing.length : Symbol(String.length, Decl(lib.es5.d.ts, --, --))
+>thing : Symbol(thing, Decl(controlFlowInstanceOfGuardPrimitives.ts, 0, 21))
+>length : Symbol(String.length, Decl(lib.es5.d.ts, --, --))
+
+    } else {
+        console.log("Aha!! It's the number " + thing.toPrecision(3));
+>console.log : Symbol(Console.log, Decl(lib.dom.d.ts, --, --))
+>console : Symbol(console, Decl(lib.dom.d.ts, --, --))
+>log : Symbol(Console.log, Decl(lib.dom.d.ts, --, --))
+>thing.toPrecision : Symbol(Number.toPrecision, Decl(lib.es5.d.ts, --, --))
+>thing : Symbol(thing, Decl(controlFlowInstanceOfGuardPrimitives.ts, 0, 21))
+>toPrecision : Symbol(Number.toPrecision, Decl(lib.es5.d.ts, --, --))
+    }
+}
+
+distinguish(new Date());
+>distinguish : Symbol(distinguish, Decl(controlFlowInstanceOfGuardPrimitives.ts, 0, 0))
+>Date : Symbol(Date, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --), Decl(lib.scripthost.d.ts, --, --))
+
+distinguish("beef");
+>distinguish : Symbol(distinguish, Decl(controlFlowInstanceOfGuardPrimitives.ts, 0, 0))
+
+distinguish(3.14159265);
+>distinguish : Symbol(distinguish, Decl(controlFlowInstanceOfGuardPrimitives.ts, 0, 0))
+

--- a/tests/baselines/reference/controlFlowInstanceOfGuardPrimitives.types
+++ b/tests/baselines/reference/controlFlowInstanceOfGuardPrimitives.types
@@ -1,0 +1,71 @@
+=== tests/cases/conformance/controlFlow/controlFlowInstanceOfGuardPrimitives.ts ===
+function distinguish(thing: string | number | Date) {
+>distinguish : (thing: string | number | Date) => void
+>thing : string | number | Date
+
+    if (thing instanceof Object) {
+>thing instanceof Object : boolean
+>thing : string | number | Date
+>Object : ObjectConstructor
+
+        console.log("Aha!! It's a Date in " + thing.getFullYear());
+>console.log("Aha!! It's a Date in " + thing.getFullYear()) : void
+>console.log : (message?: any, ...optionalParams: any[]) => void
+>console : Console
+>log : (message?: any, ...optionalParams: any[]) => void
+>"Aha!! It's a Date in " + thing.getFullYear() : string
+>"Aha!! It's a Date in " : "Aha!! It's a Date in "
+>thing.getFullYear() : number
+>thing.getFullYear : () => number
+>thing : Date
+>getFullYear : () => number
+
+    } else if (typeof thing === 'string') {
+>typeof thing === 'string' : boolean
+>typeof thing : "string" | "number" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>thing : string | number
+>'string' : "string"
+
+        console.log("Aha!! It's a string of length " + thing.length);
+>console.log("Aha!! It's a string of length " + thing.length) : void
+>console.log : (message?: any, ...optionalParams: any[]) => void
+>console : Console
+>log : (message?: any, ...optionalParams: any[]) => void
+>"Aha!! It's a string of length " + thing.length : string
+>"Aha!! It's a string of length " : "Aha!! It's a string of length "
+>thing.length : number
+>thing : string
+>length : number
+
+    } else {
+        console.log("Aha!! It's the number " + thing.toPrecision(3));
+>console.log("Aha!! It's the number " + thing.toPrecision(3)) : void
+>console.log : (message?: any, ...optionalParams: any[]) => void
+>console : Console
+>log : (message?: any, ...optionalParams: any[]) => void
+>"Aha!! It's the number " + thing.toPrecision(3) : string
+>"Aha!! It's the number " : "Aha!! It's the number "
+>thing.toPrecision(3) : string
+>thing.toPrecision : (precision?: number) => string
+>thing : number
+>toPrecision : (precision?: number) => string
+>3 : 3
+    }
+}
+
+distinguish(new Date());
+>distinguish(new Date()) : void
+>distinguish : (thing: string | number | Date) => void
+>new Date() : Date
+>Date : DateConstructor
+
+distinguish("beef");
+>distinguish("beef") : void
+>distinguish : (thing: string | number | Date) => void
+>"beef" : "beef"
+
+distinguish(3.14159265);
+>distinguish(3.14159265) : void
+>distinguish : (thing: string | number | Date) => void
+>3.14159265 : 3.14159265
+

--- a/tests/cases/conformance/controlFlow/controlFlowInstanceOfGuardPrimitives.ts
+++ b/tests/cases/conformance/controlFlow/controlFlowInstanceOfGuardPrimitives.ts
@@ -1,0 +1,13 @@
+function distinguish(thing: string | number | Date) {
+    if (thing instanceof Object) {
+        console.log("Aha!! It's a Date in " + thing.getFullYear());
+    } else if (typeof thing === 'string') {
+        console.log("Aha!! It's a string of length " + thing.length);
+    } else {
+        console.log("Aha!! It's the number " + thing.toPrecision(3));
+    }
+}
+
+distinguish(new Date());
+distinguish("beef");
+distinguish(3.14159265);

